lib/router/router.js

```diff
@@ -625,10 +632,32 @@ function updateURL(transition, state/*, inputUrl*/) {
     params.queryParams = transition._visibleQueryParams || state.queryParams;
     var url = router.recognizer.generate(handlerName, params);
 
-    if (urlMethod === 'replace') {
+    var initial   = transition.isCausedByInitialTransition;
+    var aborting  = transition.isCausedByAbortingTransition;
+
+    if (initial) {
+      // transitions during the initial transition must always use replaceURL.
+      // When the app boots, you are at a url, e.g. /foo. If some handler
+      // redirects to bar as part of the initial transition, you don't want to
+      // add a history entry for /foo. If you do, pressing back will immediately
+      // hit the redirect again and take you back to /bar, thus killing the back
+      // button
       router.replaceURL(url);
     } else {
-      router.updateURL(url);
+      if (urlMethod === 'replace' && !aborting) {
```

Comment:
Unsure how others will feel about this, but I'd prefer doing:

```js
if (initial) {
} else if (urlMethod === 'replace' && !aborting) {
} else {
}
```
Instead of the nesting you currently have.